### ISSUE TITLE
Support for data-spec/or with encoding

### ISSUE THREAD

**Issue by alzadude** [metosin/spec-tools#147]
Hi there,

Is `data-spec/or` supported with encoding? I see an example usage of `or` in the README. I am seeing some interesting results when I try and perform encoding.

For example, given the following example data specs:
```
(def sub-widget-a-spec
  (std/spec
   {:name ::sub-widget-a-spec
    :spec {:a-prop pos-int?}}))

(def sub-widget-b-spec
  (std/spec
   {:name ::sub-widget-b-spec
    :spec {:b-prop pos-int?}}))

(def sub-widget-c-spec
  (std/spec
   {:name ::sub-widget-c-spec
    :spec {:c-prop pos-int?}}))

(def widget-spec
  (std/spec
   {:name ::widget-spec
    :spec {:sub-widgets [(std/or 
                          {:sub-widget-a sub-widget-a-spec
                           :sub-widget-b sub-widget-b-spec
                           :sub-widget-c sub-widget-c-spec})]}}))
```
If I try and generate some sample data using the data spec, all is good and I see the `data-spec/or` rules being adhered to:
```
(gen/generate (s/gen widget-spec))
```
```
{:sub-widgets [{:c-prop 172831} {:a-prop 1888} {:c-prop 1886} {:c-prop 160805837} {:c-prop 1} {:a-prop 31141} {:c-prop 2} {:c-prop 1}]}
```
However, if I then try to perform encoding, for example encoding to strings, I see inconsistent results - some data in the sequence is encoded correctly, but some is left as-is:
```
(st/encode widget-spec (gen/generate (s/gen widget-spec)) st/string-transformer)
```
```
{:sub-widgets [{:b-prop 228404} {:c-prop 766} {:c-prop 1} {:a-prop "1573"} {:c-prop 33686636}]}
```
On closer inspection, it seems to always be the case that data corresponding to the first form of `data-spec/or` is correctly encoded, but data corresponding to the other forms is not.

I was wondering whether this should be expected to work or not? (since `gen/generate` works, it's just the encoding that doesn't and is therefore possibly a bug?)

Thanks!

**Comment by alzadude** [comment]
Another observation, encoding seems to work correctly for `data-spec/or` when the forms are simple spec predicates. Specifically, the issue only seems to surface when the forms are nested data specs or nested maps as used above.

For example, encoding works ok for this data spec:
```
(def widget-spec
  (std/spec
   {:name ::widget-spec
    :spec {:sub-widgets [(std/or
                          {:sub-widget-a int?
                           :sub-widget-b double?})]}}))
```

**Comment by alzadude** [comment]
I managed to track down the cause of the issue...

It seems that the current logic in `Spec/conform*` for choosing whether to return `::s/invalid` won't work for `or` specs... specifically, this line:
```
(or (and encode? (s/invalid? conformed) transformed) conformed)
```
After adding some additional logging, I discovered that spec uses the `::s/invalid` result when attempting to conform `or` specs. If a given form of the `or` returns `::s/invalid`, spec uses this result to try the next form in the or.

However the above logic prevents spec from trying the next form, because when encoding, if the result of conform was `::s/invalid`, `transformed` will be returned instead. This explains the inconsistent results for the encoding.

I managed to "fix" the issue by changing the logic as follows:
```
(or (and encode? (s/invalid? conformed) (not= x transformed) transformed) conformed)
```
As you can see, I've added an additional clause to the `and` form, so that `transformed` is only preferred if the transformer output is different to the input. The effect of this change is that `::s/invalid` will be returned to spec when non-matching forms in the `or` spec are tried.

I thought this would be enough and I was going to raise a PR, but when I run the test suite, a few tests fail in `encode-decode-test` because there are some tests that expect encoding to be idempotent (as in, encoding with the result of a previous encode leaves the data unchanged). With my code change, that wouldn't be possible since `::s/invalid` is returned for the subsequent encode.

So, I'm a bit stuck. Is idempotence important here? If it isn't, I could raise a PR. If it is, any suggestions for an alternative solution?

**Comment by alzadude** [comment]
Fixed by #148.
